Wav2Vec2ProcessorWithLM can return N best hypotheses now

vsokolovskii opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes #22150 , now the user can specify the number of hypotheses which will be returned after the decoding stage. If the specified number is higher than the actual number of hypotheses then all hypotheses will be returned. This is useful when the user wants to run the rescoring on the n-best hypotheses (check out the motivation in the linked issue). Wav2Vec2DecoderWithLMOutput class was already prepared for this feature and [this comment in the code](https://github.com/huggingface/transformers/blob/2355e463955a5392c1acf1964d89747e8b146a6f/src/transformers/models/wav2vec2_with_lm/processing_wav2vec2_with_lm.py#L571) said that this feature will be eventually added, so here it is.

I tried not to break anything that relies on the current version of the decode function, the doc string is updated with a new parameter. All tests passed. The code was well-formatted.


## Before submitting
- [ x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests? 
**Is this necessary for a such small feature?**

@younesbelkada @ArthurZucker  @sanchit-gandhi , does this make sense to you, guys? Is there anything else I should add?

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @sgugger

Integrations:

- deepspeed: HF Trainer: @stas00, Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam

Documentation: @sgugger, @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: @sgugger
- TensorFlow: @Rocketknight1

 -->

Commit:
Wav2Vec2ProcessorWithLM can return N best hypotheses now

Signed-off-by: Vladislav Sokolovskii <vladislav@parrothq.com>

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_22235). All of your documentation changes will be reflected on that endpoint.

ArthurZucker reviewed (approved):
LGTM! Thanks for working on this 😉

amyeroberts commented on src/transformers/models/wav2vec2_with_lm/processing_wav2vec2_with_lm.py:
```diff
@@ -564,17 +570,37 @@ def decode(
+                "N-best size is larger than the number of generated hypotheses, all hypotheses will be returned."
+            )
+
+        if n_best == 1:
+            return Wav2Vec2DecoderWithLMOutput(
+                text=hypotheses[0],
+                logit_score=logit_scores[0],
+                lm_score=lm_scores[0],
+                word_offsets=word_offsets[0] if word_offsets is not None else None,
+            )
+        else:
+            return Wav2Vec2DecoderWithLMOutput(
+                text=hypotheses[:n_best],
+                logit_score=logit_scores[:n_best],
+                lm_score=lm_scores[:n_best],
+                word_offsets=word_offsets[:n_best] if word_offsets is not None else None,
+            )
```
There's no need to have separate logic here - the slicing will select the 0'th element is `n_best==1` 

```suggestion
        return Wav2Vec2DecoderWithLMOutput(
            text=hypotheses[:n_best],
            logit_score=logit_scores[:n_best],
            lm_score=lm_scores[:n_best],
            word_offsets=word_offsets[:n_best] if word_offsets is not None else None,
        )
```

amyeroberts replied:
That's a good point! Sounds good to me :)

amyeroberts commented on src/transformers/models/wav2vec2_with_lm/processing_wav2vec2_with_lm.py:
```diff
@@ -450,6 +450,7 @@ def decode(
         unk_score_offset: Optional[float] = None,
         lm_score_boundary: Optional[bool] = None,
         output_word_offsets: bool = False,
+        n_best: Optional[int] = 1,
```
As `n_best` cannot be `None` 

```suggestion
        n_best: int = 1,
```

amyeroberts reviewed (approved):
LGTM - thanks for the addition! 

It would be good to have the opinion of @patrickvonplaten before merging to make sure this is in line with the decoding design

vsokolovskii committed:
Wav2Vec2ProcessorWithLM n_best cannot be None

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

sgugger commented:
We might have more luck with @sanchit-gandhi ;-)

patrickvonplaten reviewed (commented):
Thanks a lot for the PR @vsokolovskii,

Just to better understand what happens now in case we decoder a batch of logits with `n_best > 1` - > will we return a list of a list of text in this case? 

Wondering if that's the API that we want - @sanchit-gandhi wdyt?

vsokolovskii commented:
> Thanks a lot for the PR @vsokolovskii,
> 
> Just to better understand what happens now in case we decoder a batch of logits with `n_best > 1` - > will we return a list of a list of text in this case?
> 
> Wondering if that's the API that we want - @sanchit-gandhi wdyt?

Take a look at the [description of the output class arguments](https://github.com/huggingface/transformers/blob/48327c57182fdade7f7797d1eaad2d166de5c55b/src/transformers/models/wav2vec2_with_lm/processing_wav2vec2_with_lm.py#L45) that you have, you've already prepared everything for this change and I just added the return statement. There should be the possibility to get more than one hypothesis from the ASR in order to rescore it with a larger model, take a look at the motivation section in the linked issue. 🤗

sanchit-gandhi reviewed (approved):
Very cool feature @vsokolovskii! Regarding @patrickvonplaten's question about batch decoding, we don't actually have the argument `n_best` for the `batch_decode` method, it's only for the single-item, `decode` method. So currently, we'd never be returning batches of n-best hypothesis.

WDYT about adding `n_best` to the `batch_decode` method as well @vsokolovskii? In this case, I think we should match the output format to generate's beam search method as `[batches * num_sequences, output_sequences]` (see https://huggingface.co/docs/transformers/internal/generation_utils#transformers.generation.BeamSearchDecoderOnlyOutput.sequences)

Commit:
Batch decoding can return  N best hypotheses now

batch_decode was extended with the same functionality as decode
function, N best hypotheses per sample can be returned

Signed-off-by: Vladislav Sokolovskii <vladislav@parrothq.com>

vsokolovskii commented:
@sanchit-gandhi aha... gotcha. Check out the new changes, please.


> Very cool feature @vsokolovskii! Regarding @patrickvonplaten's question about batch decoding, we don't actually have the argument `n_best` for the `batch_decode` method, it's only for the single-item, `decode` method. So currently, we'd never be returning batches of n-best hypothesis.
> 
> WDYT about adding `n_best` to the `batch_decode` method as well @vsokolovskii? In this case, I think we should match the output format to generate's beam search method as `[batches * num_sequences, output_sequences]` (see https://huggingface.co/docs/transformers/internal/generation_utils#transformers.generation.BeamSearchDecoderOnlyOutput.sequences)

vsokolovskii commented:
@ArthurZucker @amyeroberts , could you please rerun the tests once the pipeline is fixed, I believe that it's not caused by my changes.

sgugger commented:
The code quality check not passing is not due to your PR at first glance, but to make sure, could you rebase on main? It has been fixed on the main branch.

vsokolovskii commented:
> The code quality check not passing is not due to your PR at first glance, but to make sure, could you rebase on main? It has been fixed on the main branch.

thanks! forgot yo update my fork
